designs/0005-eigen_expressions.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: eigen_expressions
+- Start Date: 2020-05-22
+- RFC PR: (leave this empty)
+- Stan Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow all functions in Stan Math to accept and return Eigen expressions. 
+
+# Motivation
+[motivation]: #motivation
+
+If all functions only accept and return matrices, every function must evaluate all operations it uses within the function. Evaluation of an operation means loading all matrices involved from memory to CPU, computing the opration and storing the result back in memory. For simple operations loading and storing take more time than computations.
+
+If we instead allow functions to accept and return expressions, the expresssions are only evaluated when strictly necessary. So more operations can be combined into a single expression, before it is evaluated. This results in less memory transfers and faster execution.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Eigen, the library we are using for matrix computations supports lazy evaluation. That means operations are not imediately evaluated. Instead each operation returns an object that represents this operation and references the arguments to the operation. If those arguments are operations themselves we can build arbitrarily complex expressions. These expressions are evaluated when assigned to a matrix (or `.eval()` is called on it). For example:
+
+```
+Eigen MatrixXd a, b;
+auto expr = a + 3 * b; //not evaluated yet
+MatrixXd c = expr; //now it evaluates
+```
+
+General functions must accept their arguments as templates. Which types are accepted and which are not can be restricted with requires, such as `require_eigen_t<T>`. Now these arguments can be either matrices as before or more general expressions. All functions must be updated to handle this.
+
+This is intended to require no changes in Stan lang and compiler. Here is an example:
+```
+matrix[5,5] A, B;
+matrix[5,5] C = 3 * A - 5 * B;
+```
+Variables `A`, `B` and `C` will remain of same type both in Stan Language and in compiled C++ model. The Expression `3 * A - 5 * B` will, however, be calculated more efficiently.
+
+## Example of generalizing a simple function
+Old implementation:
+```
+template <typename T1, typename T2, int R, int C>
+inline Eigen::Matrix<return_type_t<T1, T2>, R, C> add(
+    const Eigen::Matrix<T1, R, C>& m1, const Eigen::Matrix<T2, R, C>& m2) {
+  check_matching_dims("add", "m1", m1, "m2", m2);
+  return m1 + m2;
+ }
+```
+Function accepting expressions, but it still returning a matrix:
+```
+template <typename Mat1, typename Mat2,
+          typename = require_all_eigen_t<Mat1, Mat2>>
+inline auto add(const Mat1& m1, const Mat2& m2) {
+  check_matching_dims("add", "m1", m1, "m2", m2);
```

Comment:
This will evaluate m1 and m2 twice I believe.

Reply:
Nope. `check_matching_dims` just checks dims - it does not evaluate.